chiron/multistate.py

```diff
@@ -0,0 +1,968 @@
+        else:
+            return len(self._thermodynamic_states)
+
+    @property
+    def n_replicas(self):
+        """The integer number of replicas (read-only)."""
+        if self._sampler_states is None:
+            return 0
+        else:
+            return len(self._sampler_states)
+
+    @property
+    def iteration(self):
+        """The integer current iteration of the simulation (read-only).
+
+        If the simulation has not been created yet, this is None.
+
+        """
+        return self._iteration
+
+    @property
+    def mcmc_moves(self):
+        """A copy of the MCMCMoves list used to propagate the simulation.
+
+        This can be set only before creation.
+
+        """
+        return copy.deepcopy(self._mcmc_moves)
+
+    @property
+    def sampler_states(self):
+        """A copy of the sampler states list at the current iteration.
+
+        This can be set only before running.
+        """
+        return copy.deepcopy(self._sampler_states)
+
+    @property
+    def is_periodic(self):
+        """Return True if system is periodic, False if not, and None if not initialized"""
+        if self._sampler_states is None:
+            return None
+        return self._thermodynamic_states[0].is_periodic
+
+    @property
+    def metadata(self):
+        """A copy of the metadata dictionary passed on creation (read-only)."""
+        return copy.deepcopy(self._metadata)
+
+    @property
+    def is_completed(self):
+        """Check if we have reached any of the stop target criteria (read-only)"""
+        return self._is_completed()
+
+    def _compute_replica_energies(self, replica_id: int) -> np.ndarray:
+        """
+        Compute the energy for the replica in every ThermodynamicState.
+
+        Parameters
+        ----------
+        replica_id : int
+            The ID of the replica to compute energies for.
+
+        Returns
+        -------
+        np.ndarray
+            Array of energies for the specified replica across all thermodynamic states.
+        """
+        import jax.numpy as jnp
+        from chiron.states import calculate_reduced_potential_at_states
+
+        log.debug(f"{self._replica_thermodynamic_states=}")
+
+        # Determine neighborhood
+        state_index = self._replica_thermodynamic_states[replica_id]
+        neighborhood = self._neighborhood(state_index)
+        log.debug(f"{neighborhood=}")
+        # Only compute energies of the sampled states over neighborhoods.
+        energy_neighborhood_states = np.zeros(len(neighborhood))
+        neighborhood_thermodynamic_states = [
+            self._thermodynamic_states[n] for n in neighborhood
+        ]
+
+        # Retrieve sampler state associated to this replica.
+        sampler_state = self._sampler_states[replica_id]
+        log.debug(f"{sampler_state=}")
+        # Compute energy for all thermodynamic states.
+        from openmmtools.states import group_by_compatibility
+
+        for energies, the_states in [
```

Comment:
This can be greatly simplified (ideally in a way that can be JITted)---we just need to compute the dense (or sparse) matrix of `(replicas, states)`  here